From: Rui Qi <qirui.001@bytedance.com>
Subject: mm/memory-failure: allow memory allocation from emergency reserves
Date: Tue, 25 Jun 2024 10:23:42 +0800

We hope that memory errors can be successfully handled quickly, using
__GFP_MEMALLOC can help us improve the success rate of processing under
memory pressure, because to_kill struct is freed very quickly, so using
__GFP_MEMALLOC will not exacerbate memory pressure for a long time, and
more memory will be freed after killed task exiting, which will also
reduce memory pressure.

Link: https://lkml.kernel.org/r/20240625022342.6158-1-qirui.001@bytedance.com
Signed-off-by: Rui Qi <qirui.001@bytedance.com>
Cc: Miaohe Lin <linmiaohe@huawei.com>
Cc: Naoya Horiguchi <nao.horiguchi@gmail.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/memory-failure.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/mm/memory-failure.c~mm-memory-failure-allow-memory-allocation-from-emergency-reserves
+++ a/mm/memory-failure.c
@@ -451,7 +451,7 @@ static void __add_to_kill(struct task_st
 {
 	struct to_kill *tk;
 
-	tk = kmalloc(sizeof(struct to_kill), GFP_ATOMIC);
+	tk = kmalloc(sizeof(struct to_kill), GFP_ATOMIC | __GFP_MEMALLOC);
 	if (!tk) {
 		pr_err("Out of memory while machine check handling\n");
 		return;
@@ -1931,7 +1931,7 @@ static int folio_set_hugetlb_hwpoison(st
 			return -EHWPOISON;
 	}
 
-	raw_hwp = kmalloc(sizeof(struct raw_hwp_page), GFP_ATOMIC);
+	raw_hwp = kmalloc(sizeof(struct raw_hwp_page), GFP_ATOMIC | __GFP_MEMALLOC);
 	if (raw_hwp) {
 		raw_hwp->page = page;
 		llist_add(&raw_hwp->node, head);
_
